tools/mascot/src/main/java/dk/alexandra/fresco/tools/mascot/cope/CopeInputter.java

```diff
@@ -0,0 +1,97 @@
+package dk.alexandra.fresco.tools.mascot.cope;
+
+import dk.alexandra.fresco.framework.network.Network;
+import dk.alexandra.fresco.framework.util.Pair;
+import dk.alexandra.fresco.framework.util.StrictBitVector;
+import dk.alexandra.fresco.tools.mascot.MascotResourcePool;
+import dk.alexandra.fresco.tools.mascot.field.FieldElement;
+import dk.alexandra.fresco.tools.mascot.mult.MultiplyRight;
+import dk.alexandra.fresco.tools.mascot.utils.FieldElementPrg;
+import dk.alexandra.fresco.tools.mascot.utils.FieldElementPrgImpl;
+import java.math.BigInteger;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+
+/**
+ * Actively-secure implementation of the inputter party's side of the Correlated Oblivious Product
+ * Evaluation (COPE) protocol.<br>
+ * COPE allows two parties, the inputter and the signer, where the inputter holds input values e1,
+ * ..., en, and the signer holds single value s to secret-shared result of s * e1, ..., s * en. <br>
+ * This side of the protocol is to be run by the inputter party. For the other side of the protocol,
+ * see {@link CopeSigner}.
+ * 
+ */
+public class CopeInputter extends CopeShared {
+
+  private final List<FieldElementPrg> leftPrgs;
+  private final List<FieldElementPrg> rightPrgs;
+  private final MultiplyRight multiplier;
```

Comment:
It seem a little weird to use the multiplier in this class as we are not actually running the multiply protocol.

Reply:
I think we should consider if the code of the Multipliers could be organized better.

Reply:
... actually, the more I look at it ... isn't the multiplier and COPE stuff basically the same? Having the two classes seems overly complex.

Reply:
Yes, that was sloppy on my part. Originally, I had intended to write a Multiply core which would then be used a helper class both by the cope classes and what are currently the multiplier classes. I just didn't get around to it. Does that refactoring sound reasonable?

Reply:
Having both versions though is necessary because COPE is stateful (it initializes the Prgs once and then uses those), whereas the Multiply protocol is stateless (since it's used to multiply different values (both for the left and the right product) each time). As a point of reference (I'll add this to the documentation as well), the Multiply protocol is used to multiply sub-protocol of Protocol 4 in the paper.